### DISCUSSION TITLE
Need to majorly refactor this

### DISCUSSION THREAD

**Discussion by togfoxy** [togfoxy/MarsLander#257] [Ideas]
script.modules is currently structured like this:
```
local Modules =  {
    thrusters = {
        id = Enum.moduleEfficientThrusters,
        name = "Efficient Thrusters",
        cost = 225,
        mass = 20,
    },
```
etc etc

Which really means this:
Modules.thrusters.id = 1
Modules.thrusters.name = "Efficient Thrusters"
etc etc

This is causing a big problem for me now and will be a thorn in our side forever because that table has no numerical indices. So? It makes a for-loop and ipairs impossible to use. Pairs will work but pairs won't return a sequence (because there is no numerical index). When you can't use a for-loop then you can't guarantee the sequence then you get out of order things like this:

![image](https://user-images.githubusercontent.com/5599266/142404172-074ae7af-3a69-49d0-830d-c49211f0ccfa.png)

:(  This is impacting the shop menu and that will be rewritten into something graphical eventually but the problem remains. At some point in the future we'll want to loop through this in a straight forward way and it won't be possible.

I think we need to fix this now before it's too late. I suggest we move away from:
Modules.thrusters.name = "Efficient Thrusters"

and migrate to:
Modules[1].name = "Efficient Thrusters"

which becomes:
Modules[moduleEfficientThrusters].name = "Efficient Thrusters"

If we can have a think about it then I'll create a new branch just for this one refactor.

**Comment by MadByteDE** [comment]
I don't really get your point here, why do we need numerical indexes? To have the items in order they just have to be sorted once using table.sort for their id or name (alphabetical) and that's it. 
I also prefer to be able to write `Modules.thrusters` to access the module instead of `Modules[enumModuleThrusters]` which is way less readable, not speaking of Modules[1] which doesnt tell you anything about what module this actually is and one have to add an extra comment to make it clear. 
I'll look into it more when I'm home.

**Reply by togfoxy** [reply]
Take a look at the screenshot above when you get a chance. How would you normally get that into the correct sequence?

**Reply by MadByteDE** [reply]
Using the module.id just like it is done in the current build. The entries will be drawn sorted by the id of the module itself.
```lua
love.graphics.printf(itemListString, 0, y + (20*module.id), SCREEN_WIDTH, "center")
```

I get that there are use cases where tables need to be numerical indexed and sorted, but the modules table is not one of them imo because there is no reason to sort it, except for drawing the shop menu and that can easily be solved like it is done in the current main branch.
I'd personally take the readability of the code over using numerical indexes and creating extra variables like `moduleEfficientThrusters` which is just a mask hiding int 1 (the id) and that also is not created in the same definition as the module itself, meaning more code pollution.

Anyway, I might be wrong so feel free to change it :wink: .

**Reply by togfoxy** [reply]
I'm not understanding. I need to sort that shop menu. I've never done that with non-numerical indexes. What do you suggest?

**Reply by MadByteDE** [reply]
Why do you need to sort it? I already posted how it will be drawn in the correct order in my last post above using module.id.
My point is that I don't see any reason why the table has to be sorted, thats all. I also don't really like having extra variables like `Enum.moduleEfficientThrusters` instead of just NOT using indexes and do `Modules.efficientThrusters`.

Anyway, If you want to use indexes for the modules, feel free to change it, no worries. :+1:

**Comment by togfoxy** [comment]
For reference:
https://www.lua.org/pil/19.3.html
